File: getBalance.ts
const url = (address: string) =>
  `https://wallet.mainnet.alephium.org/addresses/${address}/balance`;

export interface alephiumBody {
  balance: string;
  balanceHint: string;
  lockedBalance: string;
  lockedBalanceHint: string;
  utxoNum: number;
}

export async function getBalance(address: string): Promise<string | Error> {
  if (!address) return new Error("Provide a valid address");

  const response = await fetch(url(address));
  const json: alephiumBody = await response.json();
  return json.balance;
}

File: go-app/src/balance/getBalance.go
package balance

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type alephiumBody struct {
	Balance           string `json:"balance"`
	BalanceHint       string `json:"balanceHint"`
	LockedBalance     string `json:"lockedBalance"`
	LockedBalanceHint string `json:"lockedBalanceHint"`
	UtxoNum           int    `json:"utxoNum"`
}

type URLGenerator func(address string) string

func CreateUrl(address string) string {
	url := fmt.Sprintf("https://wallet.mainnet.alephium.org/addresses/%s/balance", address)
	return url
}

func GetBalance(address string, generateURL URLGenerator) (string, error) {
	url := generateURL(address)

	resp, err := http.Get(url)
	if err != nil {
		return "", fmt.Errorf("failed to make HTTP request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("received non-200 HTTP status: %s", resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response body: %w", err)
	}

	var result alephiumBody
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("failed to unmarshal JSON: %w", err)
	}

	return result.BalanceHint, nil
}
